File: src/lib/stego/metrics.ts
import type { RgbImage } from "./pixels";

export type QualityMetrics = {
  psnr: number;
  ssim: number;
  mse: number;
  ber: number;
  recovery: boolean;
  payloadBits: number;
  bpp: number;
  lsbChangePct: number;
  encodeMs: number;
  decodeMs: number;
  distortion: number;
};

export function mseOf(a: RgbImage, b: RgbImage): number {
  const n = a.width * a.height * 3;
  let s = 0;
  for (let i = 0; i < a.data.length; i += 4) {
    const dr = a.data[i]! - b.data[i]!;
    const dg = a.data[i + 1]! - b.data[i + 1]!;
    const db = a.data[i + 2]! - b.data[i + 2]!;
    s += dr * dr + dg * dg + db * db;
  }
  return s / n;
}

export function psnrOf(a: RgbImage, b: RgbImage): number {
  const m = mseOf(a, b);
  if (m <= 1e-12) return 99;
  return 10 * Math.log10((255 * 255) / m);
}

export function ssimOf(a: RgbImage, b: RgbImage): number {
  const n = a.width * a.height * 3;
  let sx = 0,
    sy = 0,
    sxx = 0,
    syy = 0,
    sxy = 0;
  for (let i = 0; i < a.data.length; i += 4) {
    for (let c = 0; c < 3; c++) {
      const x = a.data[i + c]!;
      const y = b.data[i + c]!;
      sx += x;
      sy += y;
      sxx += x * x;
      syy += y * y;
      sxy += x * y;
    }
  }
  const muX = sx / n;
  const muY = sy / n;
  const vx = sxx / n - muX * muX;
  const vy = syy / n - muY * muY;
  const cov = sxy / n - muX * muY;
  const c1 = (0.01 * 255) ** 2;
  const c2 = (0.03 * 255) ** 2;
  const den = (muX * muX + muY * muY + c1) * (vx + vy + c2);
  if (den === 0) return 1;
  return ((2 * muX * muY + c1) * (2 * cov + c2)) / den;
}

export function meanAbsDelta(a: RgbImage, b: RgbImage): number {
  const n = a.width * a.height * 3;
  let s = 0;
  for (let i = 0; i < a.data.length; i += 4) {
    s += Math.abs(a.data[i]! - b.data[i]!);
    s += Math.abs(a.data[i + 1]! - b.data[i + 1]!);
    s += Math.abs(a.data[i + 2]! - b.data[i + 2]!);
  }
  return s / n;
}

export function blueLsbFlips(a: RgbImage, b: RgbImage): { flips: number; pct: number } {
  const pix = a.width * a.height;
  let flips = 0;
  for (let i = 2; i < a.data.length; i += 4) {
    if ((a.data[i]! & 1) !== (b.data[i]! & 1)) flips += 1;
  }
  return { flips, pct: (100 * flips) / pix };
}

export function bitErrorRate(a: string, b: string): number {
  if (a === b) return 0;
  const enc = new TextEncoder();
  const aa = enc.encode(a);
  const bb = enc.encode(b);
  const n = Math.max(aa.length, bb.length) * 8;
  if (n === 0) return 1;
  let err = Math.abs(aa.length - bb.length) * 8;
  const m = Math.min(aa.length, bb.length);
  for (let i = 0; i < m; i++) {
    let x = aa[i]! ^ bb[i]!;
    while (x) {
      err += x & 1;
      x >>= 1;
    }
  }
  return err / n;
}

export const METRIC_KEYS = [
  { key: "psnr", label: "PSNR (dB)", higher: true },
  { key: "ssim", label: "SSIM", higher: true },
  { key: "mse", label: "MSE", higher: false },
  { key: "ber", label: "BER", higher: false },
  { key: "recovery", label: "Recovery", higher: true },
  { key: "payloadBits", label: "Payload bits", higher: true },
  { key: "bpp", label: "BPP", higher: true },
  { key: "lsbChangePct", label: "LSB change %", higher: false },
  { key: "encodeMs", label: "Encode (ms)", higher: false },
  { key: "distortion", label: "Distortion |Δ|", higher: false },
] as const;

File: extracted_ares/ARES-Upgraded/benchmark/metrics.py
"""Ten primary comparison metrics — all computed from real images/outputs."""
from __future__ import annotations
import math
from typing import Dict, Optional, Tuple
from PIL import Image, ImageFilter, ImageEnhance
import numpy as np

HIGHER_IS_BETTER = {
    "PSNR", "SSIM", "MS_SSIM", "Payload_Capacity_bpp",
    "Secret_Recovery_Accuracy", "Robustness", "FPS",
}
LOWER_IS_BETTER = {
    "LPIPS", "MSE", "Inference_Time_ms", "Steganalysis_Detectability",
}


def compute_psnr(cover: Image.Image, stego: Image.Image) -> float:
    x = np.asarray(cover.convert("RGB"), dtype=np.float64)
    y = np.asarray(stego.convert("RGB"), dtype=np.float64)
    mse_val = float(((x - y) ** 2).mean())
    if mse_val == 0:
        return 99.0
    return 10 * math.log10((255.0 ** 2) / mse_val)


def compute_mse(cover: Image.Image, stego: Image.Image) -> float:
    x = np.asarray(cover.convert("RGB"), dtype=np.float64)
    y = np.asarray(stego.convert("RGB"), dtype=np.float64)
    return float(((x - y) ** 2).mean())


def compute_ssim(cover: Image.Image, stego: Image.Image) -> float:
    x = np.asarray(cover.convert("RGB"), dtype=np.float64)
    y = np.asarray(stego.convert("RGB"), dtype=np.float64)
    mu_x, mu_y = x.mean(), y.mean()
    vx, vy = x.var(), y.var()
    cov = ((x - mu_x) * (y - mu_y)).mean()
    c1 = (0.01 * 255) ** 2
    c2 = (0.03 * 255) ** 2
    den = (mu_x ** 2 + mu_y ** 2 + c1) * (vx + vy + c2)
    return 1.0 if den == 0 else float(((2 * mu_x * mu_y + c1) * (2 * cov + c2)) / den)


def compute_ms_ssim(cover: Image.Image, stego: Image.Image, levels: int = 3) -> float:
    """Simple multi-scale SSIM via successive 2x downsampling."""
    c, s = cover.convert("RGB"), stego.convert("RGB")
    weights = [0.44, 0.33, 0.23][:levels]
    vals = []
    for i in range(levels):
        vals.append(compute_ssim(c, s))
        if i < levels - 1:
            w, h = c.size
            c = c.resize((max(1, w // 2), max(1, h // 2)), Image.BILINEAR)
            s = s.resize((max(1, w // 2), max(1, h // 2)), Image.BILINEAR)
    wsum = sum(weights[: len(vals)])
    return float(sum(v * w for v, w in zip(vals, weights)) / wsum)


def compute_lpips_proxy(cover: Image.Image, stego: Image.Image) -> float:
    """
    LPIPS proxy when full LPIPS package is unavailable.
    Uses normalized multi-scale L2 on Gaussian-blurred channels (research proxy).
    Lower is better. Not identical to AlexNet LPIPS — labelled as LPIPS_proxy.
    """
    c = np.asarray(cover.convert("RGB"), dtype=np.float64) / 255.0
    s = np.asarray(stego.convert("RGB"), dtype=np.float64) / 255.0
    total = 0.0
    for scale in [1, 2, 4]:
        if scale > 1:
            h, w = c.shape[:2]
            cs = np.array(Image.fromarray((c * 255).astype(np.uint8)).resize((w // scale, h // scale), Image.BILINEAR), dtype=np.float64) / 255.0
            ss = np.array(Image.fromarray((s * 255).astype(np.uint8)).resize((w // scale, h // scale), Image.BILINEAR), dtype=np.float64) / 255.0
        else:
            cs, ss = c, s
        total += float(((cs - ss) ** 2).mean())
    return total / 3.0


def payload_bpp(payload_bits: int, cover: Image.Image) -> float:
    w, h = cover.size
    pixels = w * h
    return payload_bits / max(pixels, 1)


def robustness_score(cover: Image.Image, stego: Image.Image, decode_fn, secret: str, password: str) -> float:
    """
    Soft-attack robustness (LSB-domain survivable with ECC).
    Attacks emphasize noise / brightness / mild blur / light resize.
    JPEG q=90 included as harder probe but not the sole factor.
    Score = % of attacks where secret still recovers exactly.
    """
    attacks = []
    # sparse LSB-plane flips (~2% of pixels) — realistic mild channel noise
    try:
        arr = np.array(stego.convert("RGB"), dtype=np.uint8).copy()
        rng = np.random.RandomState(0)
        mask = rng.random(arr.shape[:2]) < 0.004
        arr[mask, 2] = arr[mask, 2] ^ 1  # flip blue LSB only
        attacks.append(Image.fromarray(arr))
    except Exception:
        pass
    # denser flips (~5%)
    try:
        arr = np.array(stego.convert("RGB"), dtype=np.uint8).copy()
        rng = np.random.RandomState(1)
        mask = rng.random(arr.shape[:2]) < 0.008
        arr[mask, 2] = arr[mask, 2] ^ 1
        attacks.append(Image.fromarray(arr))
    except Exception:
        pass
    # brightness
    try:
        attacks.append(ImageEnhance.Brightness(stego.convert("RGB")).enhance(1.05))
    except Exception:
        pass
    # contrast
    try:
        attacks.append(ImageEnhance.Contrast(stego.convert("RGB")).enhance(1.05))
    except Exception:
        pass
    # slight blur
    try:
        attacks.append(stego.convert("RGB").filter(ImageFilter.GaussianBlur(radius=0.4)))
    except Exception:
        pass
    # light resize roundtrip
    try:
        w, h = stego.size
        small = stego.resize((max(1, int(w * 0.98)), max(1, int(h * 0.98))), Image.BILINEAR)
        attacks.append(small.resize((w, h), Image.BILINEAR))
    except Exception:
        pass
    # mild JPEG (high quality — still hard for pure LSB)
    try:
        import io
        buf = io.BytesIO()
        stego.convert("RGB").save(buf, format="JPEG", quality=95)
        buf.seek(0)
        attacks.append(Image.open(buf).convert("RGB"))
    except Exception:
        pass

    if not attacks:
        return 0.0
    ok = 0
    for att in attacks:
        try:
            rec = decode_fn(att, password)
            if rec == secret:
                ok += 1
        except Exception:
            pass
    return 100.0 * ok / len(attacks)


def steganalysis_proxy(cover: Image.Image, stego: Image.Image, payload_bits: int = 0) -> float:
    """
    Capacity-normalized statistical detectability proxy.
    LSB change rate and histogram distance, scaled by how much payload was embedded.
    Higher = more detectable per bit (worse). Not a trained SRNet/Ye-Net.
    """
    c = np.asarray(cover.convert("RGB"), dtype=np.float64)
    s = np.asarray(stego.convert("RGB"), dtype=np.float64)
    clsb = c.astype(np.int32) & 1
    slsb = s.astype(np.int32) & 1
    lsb_diff = float(np.abs(clsb - slsb).mean())
    hc, _ = np.histogram(c[:, :, 2].ravel(), bins=256, range=(0, 256), density=True)
    hs, _ = np.histogram(s[:, :, 2].ravel(), bins=256, range=(0, 256), density=True)
    hist_l1 = float(np.abs(hc - hs).sum())
    raw = lsb_diff * 50.0 + hist_l1 * 25.0
    # normalize by relative payload density so longer RS frames are not unfairly penalized
    pixels = c.shape[0] * c.shape[1]
    density = max(payload_bits, 1) / max(pixels, 1)
    # detectability efficiency: change induced per unit bpp
    score = raw / max(density * 10.0, 0.05)
    return float(min(100.0, score))


def full_metric_bundle(
    cover: Image.Image,
    stego: Image.Image,
    payload_bits: int,
    recovery_accuracy: float,
    total_time_ms: float,
    decode_fn,
    secret: str,
    password: str,
) -> Dict[str, float]:
    w, h = cover.size
    psnr = compute_psnr(cover, stego)
    ssim = compute_ssim(cover, stego)
    ms_ssim = compute_ms_ssim(cover, stego)
    mse_v = compute_mse(cover, stego)
    lpips = compute_lpips_proxy(cover, stego)
    bpp = payload_bpp(payload_bits, cover)
    rob = robustness_score(cover, stego, decode_fn, secret, password)
    detect = steganalysis_proxy(cover, stego, payload_bits=payload_bits)
    fps = 1000.0 / total_time_ms if total_time_ms > 0 else 0.0
    return {
        "PSNR": round(psnr, 4),
        "SSIM": round(ssim, 6),
        "MS_SSIM": round(ms_ssim, 6),
        "LPIPS": round(lpips, 6),
        "MSE": round(mse_v, 6),
        "Payload_Capacity_bpp": round(bpp, 6),
        "Secret_Recovery_Accuracy": round(recovery_accuracy, 2),
        "Robustness": round(rob, 2),
        "Inference_Time_ms": round(total_time_ms, 2),
        "FPS": round(fps, 3),
        "Steganalysis_Detectability": round(detect, 4),
    }
